## Solutions to problem_1 of Assignment_1

In [1]:
import pandas as pd
from scipy.stats import multivariate_normal
# for displaying only first 10 rows and first 10 columns
pd.set_option('display.max_rows', 10)
pd.set_option('display.max_columns', 10)

By default, `pd.read_csv` uses `header=0` (when the names parameter is also not specified) which means the first (i.e. 0th-indexed) line is interpreted as column names.

If our data has no header, then use
`pd.read_csv(..., header=None)`

In [2]:
data_train = pd.read_csv("P1_data/P1_data_train.csv", header=None)
label_train = pd.read_csv("P1_data/P1_labels_train.csv", header=None)

Displaying first 5 labels using `Dataframe.head()`

In [3]:
label_train.head()

0
0  6
1  5
2  6
3  5
4  6

Creating two list to store the index values of 5 and 6
`index_5` & `index_6` are list which contains row no of label 5 & 6 respectively.

In [4]:
index_5 = []
index_6 = []
for i, j in label_train.iterrows():
    if j[0] == 5:
        index_5.append(i)
    else:
        index_6.append(i)

No. of elements which are labelled as 5 in training dataset.

In [5]:
len(index_5)

396

### Individual probability of label being 5 or 6
i.e. `prior probabilities` <br> `index_5_data` and `index_6_data` stores data which are labelled as 5 and 6 respectively.

In [6]:
prob_5 = len(index_5)/len(label_train)
prob_6 = 1 - prob_5
index_5_data = data_train.iloc[index_5]
index_6_data = data_train.iloc[index_6]

### Mean of label 5

In [7]:
mean_5 = index_5_data.mean()
mean_5

0      5.000000
1      5.083333
2      8.515152
3     10.340909
4     10.772727
        ...    
59    10.563131
60     9.065657
61     6.739899
62     5.068182
63     5.244949
Length: 64, dtype: float64

### Meanof label 6

In [8]:
mean_6 = index_6_data.mean()
mean_6

0      5.196850
1      5.089239
2      5.952756
3     10.070866
4      8.569554
        ...    
59     9.440945
60    11.301837
61     9.955381
62     6.664042
63     5.291339
Length: 64, dtype: float64

### Covariance of label 5

In [9]:
covariance_5 = index_5_data.cov()
covariance_5

0          1          2          3          4     ...      \
0   38.977215  21.539241  23.653165  20.437975  19.296203    ...       
1   21.539241  38.603165  23.070886  22.161392  18.555696    ...       
2   23.653165  23.070886  44.787112  26.996087  24.436364    ...       
3   20.437975  22.161392  26.996087  40.984753  21.085270    ...       
4   19.296203  18.555696  24.436364  21.085270  35.897583    ...       
..        ...        ...        ...        ...        ...    ...       
59  20.951899  21.714979  25.780054  27.597411  22.561220    ...       
60  24.010127  21.029958  25.229382  28.726928  20.278251    ...       
61  24.774684  27.922996  24.033065  27.073705  17.697699    ...       
62  24.913924  23.622152  19.397699  19.779229  15.934522    ...       
63  25.784810  25.589662  26.969697  22.101093  22.858343    ...       

           59         60         61         62         63  
0   20.951899  24.010127  24.774684  24.913924  25.784810  
1   21.714979  21.029958  27.922996  23.622152  25.589662  
2   25.780054  25.229382  24.033065  19.397699  26.969697  
3   27.597411  28.726928  27.073705  19.779229  22.101093  
4   22.561220  20.278251  17.697699  15.934522  22.858343  
..        ...        ...        ...        ...        ...  
59  40.368156  28.373060  23.967095  20.870368  22.292092  
60  28.373060  45.772893  30.470285  24.362601  22.201598  
61  23.967095  30.470285  45.742303  26.319045  24.555012  
62  20.870368  24.362601  26.319045  39.385213  25.016168  
63  22.292092  22.201598  24.555012  25.016168  41.187949  

[64 rows x 64 columns]

### Covariance of label 6

In [10]:
covariance_6 = index_6_data.cov()
covariance_6

0          1          2          3          4     ...      \
0   43.037464  23.490282  25.693535  18.425487  22.721799    ...       
1   23.490282  40.402542  26.048964  23.064712  21.730619    ...       
2   25.693535  26.048964  42.903025  24.382304  23.308558    ...       
3   18.425487  23.064712  24.382304  40.897596  17.604269    ...       
4   22.721799  21.730619  23.308558  17.604269  44.424755    ...       
..        ...        ...        ...        ...        ...    ...       
59  24.420866  23.131600  26.439308  26.029196  21.556092    ...       
60  21.703585  18.741415  20.622192  23.512764  16.627635    ...       
61  20.914070  25.182940  22.437360  26.100539  19.509691    ...       
62  24.076834  26.837954  20.970928  19.950186  20.189218    ...       
63  28.568815  27.823933  29.234853  21.013510  25.773104    ...       

           59         60         61         62         63  
0   24.420866  21.703585  20.914070  24.076834  28.568815  
1   23.131600  18.741415  25.182940  26.837954  27.823933  
2   26.439308  20.622192  22.437360  20.970928  29.234853  
3   26.029196  23.512764  26.100539  19.950186  21.013510  
4   21.556092  16.627635  19.509691  20.189218  25.773104  
..        ...        ...        ...        ...        ...  
59  43.278740  28.656030  27.669726  25.219581  25.708040  
60  28.656030  37.511286  25.802977  22.517461  20.551305  
61  27.669726  25.802977  40.969056  26.766549  22.934086  
62  25.219581  22.517461  26.766549  46.107888  26.600767  
63  25.708040  20.551305  22.934086  26.600767  42.154372  

[64 rows x 64 columns]

Continuing with the test data

In [11]:
data_test = pd.read_csv("P1_data/P1_data_test.csv", header=None)
data_test

0   1   2   3   4  ...  59  60  61  62  63
0     4   3  16  16  11 ...   9  12   0   1   5
1    15   0  11   0  16 ...   4  15   0   0   0
2     0   0   0   8   9 ...   0   0   0   0   0
3     3  10  16  15  16 ...  16  15  12   4  15
4     0   0   0   0   0 ...   0   0   0   0   0
..   ..  ..  ..  ..  .. ...  ..  ..  ..  ..  ..
328   0   0   0   4   0 ...   0   1   6   0   0
329   0   0   0   8   2 ...  10   6   5   0   0
330  16  14  16  16  16 ...  16  16  10   9  16
331   0   0   0   0  10 ...   0   0   0   0   0
332   5   0  16  16   8 ...  16  16  16  15   8

[333 rows x 64 columns]

Making multivariate_normal pdf of test data given `mean` and `variance` calculated from training data. <br> And calculating posterior probabilties by multiplying with corresponding prior probabilities.

In [12]:
y1 = multivariate_normal.pdf(data_test,mean=mean_5,cov=covariance_5)*prob_5
y2 = multivariate_normal.pdf(data_test,mean=mean_6,cov=covariance_6)*prob_6

In [13]:
li = []
for i in range(len(y1)):
    if y1[i] > y2[i]:
        li.append(5)
    else:
        li.append(6)

In [14]:
true_label = pd.read_csv('P1_data/P1_labels_test.csv', header = None)

In [15]:
count = 0
for i in range(len(true_label)):
    if(true_label[0][i] == li[i]):
        count += 1
        
count

257

It gave correct label of 257 data out of 333. <br>
Now, finding the confusion matrix as we know the correct label and the predicted label.

In [16]:
from sklearn.metrics import confusion_matrix
conf_mat = confusion_matrix(true_label,li)
print(conf_mat)

[[106  49]
 [ 27 151]]


As the misclassification error is found to be 22.8%. Accuracy is 77.2%.

In [17]:
misclassification_error = (conf_mat[0][1]+conf_mat[1][0])/(conf_mat[0][0]+conf_mat[0][1]+conf_mat[1][0]+conf_mat[1][1])
misclassification_error

0.22822822822822822